connect/tunnel_api/tunnel_api.go

```diff
@@ -0,0 +1,853 @@
+	openvpnConfTemplate, err := template.New("config").Parse(openvpnClientConfTemplate)
+	if err != nil {
+		return err
+	}
+
+	dstService := &corev1.Service{}
+
+	for i := 0; i < 10; i++ {
+		err = tunnel.DstClient.Get(context.TODO(), client.ObjectKey{Name: serviceName, Namespace: tunnel.Options.Namespace}, dstService)
+		if err != nil {
+			return err
+		}
+		if dstService.Status.LoadBalancer.Ingress != nil {
+			break
+		}
+		time.Sleep(time.Second * 3)
+	}
+
+	configdata := openvpnConfigData{
+		Port:     strconv.Itoa(int(tunnel.Options.ServerPort)),
+		CA:       tunnel.Options.CACrt.String(),
+		Key:      tunnel.Options.ClientKey.String(),
+		Crt:      tunnel.Options.ClientCrt.String(),
+		Endpoint: dstService.Status.LoadBalancer.Ingress[0].Hostname,
+	}
+
+	err = openvpnConfTemplate.Execute(&openvpnConf, configdata)
+	if err != nil {
+		return err
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: *&tunnel.Options.Namespace,
+		},
+		Data: map[string][]byte{
+			"openvpn.conf": openvpnConf.Bytes(),
+		},
+	}
+
+	mode := int32(0400)
+	volumes := []v1.Volume{
+		{
+			Name: serviceConfig,
+			VolumeSource: v1.VolumeSource{
+				Secret: &v1.SecretVolumeSource{
+					DefaultMode: &mode,
+					SecretName:  serviceName,
+				},
+			},
+		},
+	}
+
+	volumeMounts := []v1.VolumeMount{
+		{
+			Name:      serviceConfig,
+			MountPath: "/openvpn.conf",
+			SubPath:   "openvpn.conf",
+		},
+	}
+
+	trueBool := true
+	containers := []corev1.Container{
+		{
+			Name:  serviceName,
+			Image: *&tunnel.Options.ServerImage,
+			Command: []string{
+				"bash",
+				"-c",
+				"mkdir /tmp/openvpn && openvpn --config /openvpn.conf",
+			},
+			VolumeMounts:    volumeMounts,
+			SecurityContext: &corev1.SecurityContext{Privileged: &trueBool},
+		},
+		{
+			Name:  "socat",
+			Image: *&tunnel.Options.ServerImage,
+			Command: []string{
+				"bash",
+				"-c",
+				"socat TCP-LISTEN:8443,fork,reuseaddr TCP:${KUBERNETES_SERVICE_HOST}:${KUBERNETES_SERVICE_PORT_HTTPS}",
+			},
+		},
+	}
+
+	rootUser := int64(0)
+	podSpec := corev1.PodSpec{
+		ServiceAccountName: serviceName,
+		SecurityContext:    &corev1.PodSecurityContext{RunAsUser: &rootUser},
+		Containers:         containers,
+		Volumes:            volumes,
+	}
+
+	deploymentMeta := metav1.ObjectMeta{
+		Name:      serviceName,
+		Namespace: *&tunnel.Options.Namespace,
+	}
+
+	deploymentSelector := &metav1.LabelSelector{
+		MatchLabels: deploymentLabels,
+	}
+
+	podTemplateSpec := v1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: deploymentLabels,
+		},
+		Spec: podSpec,
+	}
+
+	err = tunnel.SrcClient.Create(context.TODO(), namespace, &client.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	err = createRBAC(tunnel, "src")
+	if err != nil {
+		return err
+	}
+	err = tunnel.SrcClient.Create(context.TODO(), secret, &client.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	if tunnel.SrcVersionMinor < 9 {
```

Comment:
I wonder if we can instead use the dynamic client and see if this version exists rather than use a minor version, especially as we near 4.9?

Reply:
This is kubernetes Minor, distinct from openshift. It won't conflict with openshift 4.9.  OpenShift 4.8 kubernetes minor is 1.21, 4.9 is probably 1.22, etc.

Reply:
I see, do you think the logic would be more "safe" checking versions/resources in a cluster vs relying on reported minor versions? 

Just a thought, and maybe this is a follow on the issue as well?

Reply:
Based on the comment below, I think a follow-up issue because we won't be able to remove this completely.

Reply:
https://github.com/konveyor/crane-lib/issues/89